File: Tree.py
from Node import Node


class Tree:
	def __init__(self):
	        self.root = None
	
	# def insert(self, data):
	# 	if(self.root == None):
	# 		self.root = Node(data)			
	# 	else:
	# 		if(self.root.left == None):
	# 			self.root.left = Node(data)
	# 		elif(self.root.right == None):
	# 			self.root.right = Node(data)
	
	def is_empty(self):
		if(self.root == None):
			return 1
		else:
			return 0
		self.root=None

	def set_root(self,node):
		self.root=node

			

	# def inOrder(self):
	# 	if self.is_empty():
	# 		return
	# 	else:
	# 		self.root.left.inOrder()
	# 		print(self.root.data)
	# 		self.root.right.inOrder()

	def get_tree(self):
		print (self.root.data)
		print (self.root.left.data)
		print (self.root.right.data)
		print (self.root.left.right.data)
		print (self.root.left.left.data)
		print (self.root.right.left.data)
		print (self.root.right.right.data)

	
	def inOrder(self,root):
		if root ==None :
			return
		else:
			#print(root.left.data)
			self.inOrder(root.left)
			print(root.data)
			self.inOrder(root.right)
 		
	def get_leafs(self,root):
		if root ==None :
			return
		else:
			if(root.is_leaf()):
				print(root.data)
			self.get_leafs(root.left)
			
			self.get_leafs(root.right)
 		
	

File: tree.ts
import {BinaryTreeNode} from './node'

export class BinaryTree {
    traversalArray:any[]=[]
    totalLeafs:any[]=[]
    root: BinaryTreeNode;

    constructor(root:BinaryTreeNode){
        this.root = root
    }
    isEmpty(){
        if(this.root == null){
            return 'Yes , This tree is empty'
        }else{
            return 'No , This tree contain values'
        }
    }

    getLeafs(treeNode:BinaryTreeNode){
        if(treeNode == null){
            this.getLeafs(this.root)  
        }
        if (treeNode == undefined) {
            return false
        }
        if(treeNode.isLeaf()){
            treeNode.isLeaf()
            console.log('leaf  >>>> ',treeNode.isLeaf())
            this.totalLeafs.push(treeNode.isLeaf())
        }else{
            if(treeNode.refLeft && !treeNode.refRight){
                this.getLeafs(treeNode.refLeft)
            }else if(!treeNode.refLeft && treeNode.refRight){
                this.getLeafs(treeNode.refRight)
            }else if(treeNode.refLeft && treeNode.refRight){
                this.getLeafs(treeNode.refLeft)
                this.getLeafs(treeNode.refRight)
            }
        }
    }

    onlyUnique(value, index, self) { 
        return self.indexOf(value) === index;
    }

    traversalInOrder(treeNode:BinaryTreeNode){
         if (treeNode == null) {
            this.traversalInOrder(this.root);
        }
        if (treeNode == undefined) {
            return false;
        }
        if( treeNode.refLeft && !treeNode.refRight ){
            this.traversalInOrder(treeNode.refLeft);
            this.traversalArray.push(treeNode.refLeft.value)
        }else if ( treeNode.refLeft && treeNode.refRight) {
            this.traversalInOrder(treeNode.refLeft);
            this.traversalInOrder(treeNode.refRight);
            this.traversalArray.push(treeNode.refLeft.value,treeNode.value,treeNode.refRight.value)
        }else if ( treeNode.refRight && !treeNode.refLeft  ){
            this.traversalInOrder(treeNode.refRight);
        }
        var unique = this.traversalArray.filter( this.onlyUnique );
        console.log('Traversal in order : ', unique)
    }
}